bundles/org.openhab.core.io.rest.core/src/main/java/org/openhab/core/io/rest/core/internal/tag/TagResource.java

```diff
@@ -74,28 +84,195 @@ public class TagResource implements RESTResource {
+        SemanticTag parentTag = null;
+        int lastSeparator = uid.lastIndexOf("_");
+        if (lastSeparator <= 0) {
+            return getTagResponse(Status.BAD_REQUEST, null, locale, "Invalid tag identifier " + uid);
+        }
+        String name = uid.substring(lastSeparator + 1);
+        parentTag = semanticTagRegistry.get(uid.substring(0, lastSeparator));
+        if (parentTag == null) {
+            return getTagResponse(Status.BAD_REQUEST, null, locale,
+                    "No existing parent tag with id " + uid.substring(0, lastSeparator));
+        } else if (!name.matches("[A-Z][a-zA-Z0-9]+")) {
+            return getTagResponse(Status.BAD_REQUEST, null, locale, "Invalid tag name " + name);
+        }
+        Class<? extends Tag> tagClass = SemanticTags.getById(name);
+        if (tagClass != null) {
+            // report a conflict
+            return getTagResponse(Status.CONFLICT, semanticTagRegistry.get(tagClass.getAnnotation(TagInfo.class).id()),
+                    locale, "Tag " + tagClass.getAnnotation(TagInfo.class).id() + " already exists!");
+        }
+
+        uid = parentTag.getUID() + "_" + name;
+        tag = new SemanticTagImpl(uid, data.label, data.description, data.synonyms);
+        managedSemanticTagProvider.add(tag);
+
+        return getTagResponse(Status.CREATED, tag, locale, null);
+    }
+
+    @DELETE
+    @RolesAllowed({ Role.ADMIN })
+    @Path("/{tagId}")
+    @Operation(operationId = "removeSemanticTag", summary = "Removes a semantic tag and its sub tags from the registry.", security = {
+            @SecurityRequirement(name = "oauth2", scopes = { "admin" }) }, responses = {
+                    @ApiResponse(responseCode = "200", description = "OK, was deleted."),
+                    @ApiResponse(responseCode = "404", description = "Custom tag not found."),
+                    @ApiResponse(responseCode = "405", description = "Custom tag not editable.") })
+    public Response remove(
+            @HeaderParam(HttpHeaders.ACCEPT_LANGUAGE) @Parameter(description = "language") @Nullable String language,
+            @PathParam("tagId") @Parameter(description = "tag id") String tagId) {
+        final Locale locale = localeService.getLocale(language);
+
+        String uid = tagId.trim();
+
+        // check whether tag exists and throw 404 if not
+        SemanticTag tag = semanticTagRegistry.get(uid);
+        if (tag == null) {
+            return getTagResponse(Status.NOT_FOUND, null, locale, "Tag " + uid + " does not exist!");
+        }
+
+        // Get tags in reverse order
+        List<String> uids = semanticTagRegistry.getSubTree(tag).stream().map(t -> t.getUID())
+                .sorted((element1, element2) -> element2.compareTo(element1)).collect(Collectors.toList());
+        for (String id : uids) {
+            // ask whether the tag exists as a managed tag, so it can get updated, 405 otherwise
+            if (managedSemanticTagProvider.get(id) == null) {
```

Comment:
Can this be simplified to a place where we just confirm given tag `id` is managed or not? I suppose we will not reach a place where parent tag might be managed and its children might not?

Reply:
This is currently not possible but could become possible later when an unmanaged provider is added.

I am not yet sure if we should or not allow unmanaged tags to be added as childs to a managed tag and vice-versa. Because this will make loading of unmanaged and managed providers very difficult.